File: alc/genast.c
// Generate code from an AST tree for the alic compiler.
// (c) 2025 Warren Toomey, GPL3

#include "alic.h"
#include "proto.h"

static void gen_IF(ASTnode * n);
static void gen_WHILE(ASTnode * n);
static void gen_local(ASTnode *n);
static int gen_funccall(ASTnode *n);

// Generate and return a new label number
static int labelid = 1;
int genlabel(void) {
  return (labelid++);
}

// Given an AST, generate assembly code recursively.
// Return the temporary id with the tree's final value.
int genAST(ASTnode * n) {
  int lefttemp, righttemp;
  int label;

  // Empty tree, do nothing
  if (n == NULL) return (NOREG);

  // Do special case nodes before the general processing
  switch (n->op) {
  case A_PRINT:
    righttemp = genAST(n->right);
    label= add_strlit(n->left->strlit);
    cgprint(label, righttemp, n->right->type); return(NOREG);
  case A_LOCAL:
    gen_local(n); return(NOREG);
  case A_FUNCCALL:
    return(gen_funccall(n));
  case A_IF:
    gen_IF(n); return(NOREG);
  case A_WHILE:
    gen_WHILE(n); return(NOREG);
  case A_FOR:
    // Generate the initial code
    genAST(n->right);

    // Now call gen_WHILE() using the left and mid children
    gen_WHILE(n); return(NOREG);
  }

  // Load the left and right sub-trees into temporaries
  if (n->left)  lefttemp  = genAST(n->left);
  if (n->right) righttemp = genAST(n->right);

  // General processing
  switch (n->op) {
  case A_NUMLIT:
    return (cgloadlit(n->litval, n->type));
  case A_ADD:
    return (cgadd(lefttemp, righttemp, n->type));
  case A_SUBTRACT:
    return (cgsub(lefttemp, righttemp, n->type));
  case A_MULTIPLY:
    return (cgmul(lefttemp, righttemp, n->type));
  case A_DIVIDE:
    return (cgdiv(lefttemp, righttemp, n->type));
  case A_NEGATE:
    return (cgnegate(lefttemp, n->type));
  case A_IDENT:
    return (cgloadvar(n->sym));
  case A_ASSIGN:
    cgstorvar(lefttemp, n->type, n->sym);
    return (NOREG);
  case A_CAST:
    return (cgcast(lefttemp, n->left->type, n->type));
  case A_EQ:
  case A_NE:
  case A_LT:
  case A_GT:
  case A_LE:
  case A_GE:
    return (cgcompare(n->op, lefttemp, righttemp, n->left->type));
  case A_INVERT:
    return (cginvert(lefttemp, n->type));
  case A_AND:
    return (cgand(lefttemp, righttemp, n->type));
  case A_OR:
    return (cgor(lefttemp, righttemp, n->type));
  case A_XOR:
    return (cgxor(lefttemp, righttemp, n->type));
  case A_LSHIFT:
    return (cgshl(lefttemp, righttemp, n->type));
  case A_RSHIFT:
    return (cgshr(lefttemp, righttemp, n->type));
  case A_NOT:
    return (cgnot(lefttemp, n->type));
  case A_GLUE:
    return (NOREG);
  }

  // Error
  fatal("genAST() unknown op %d\n", n->op);
  return (NOREG);
}

// Generate the code for an IF statement
// and an optional ELSE clause.
static void gen_IF(ASTnode * n) {
  int Lfalse, Lend = 0;
  int t1;

  // Generate two labels: one for the
  // false compound statement, and one
  // for the end of the overall IF statement.
  // When there is no ELSE clause, Lfalse
  // _is_ the ending label!
  Lfalse = genlabel();
  if (n->right)
    Lend = genlabel();

  // Generate the condition code
  t1 = genAST(n->left);

  // Jump if false to the false label
  cgjump_if_false(t1, Lfalse);

  // Generate the true statement block
  genAST(n->mid);

  // If there is an optional ELSE clause,
  // generate the jump to skip to the end
  if (n->right) {
    // QBE doesn't like two jump instructions in a row, and
    // a break at the end of a true IF section causes this.
    // The solution is to insert a label before the IF jump.
    cglabel(genlabel());
    cgjump(Lend);
  }

  // Now the false label
  cglabel(Lfalse);

  // Optional ELSE clause: generate the false
  // statement block and the end label
  if (n->right) {
    genAST(n->right);
    cglabel(Lend);
  }
}

// Generate the code for a WHILE statement
static void gen_WHILE(ASTnode * n) {
  int Lstart, Lend;
  int t1;

  // Generate the start and end labels
  // and output the start label
  Lstart = genlabel();
  Lend = genlabel();
  cglabel(Lstart);

  // Generate the condition code
  t1 = genAST(n->left);

  // Jump if false to the end label
  cgjump_if_false(t1, Lend);

  // Generate the statement block for the WHILE body
  genAST(n->mid);

  // Finally output the jump back to the condition,
  // and the end label
  cgjump(Lstart);
  cglabel(Lend);
}

// Generate space for a local variable
// and assign its value
void gen_local(ASTnode *n) {
  int lefttemp;

  // Allocate space for the variable
  cgaddlocal(n->type, n->sym);

  // Get the expression's value on the left
  lefttemp  = genAST(n->left);

  // Store this into the local variable
  cgstorvar(lefttemp, n->type, n->sym);

  // and generate any code for the other children
  genAST(n->mid);
  genAST(n->right);
}

// Generate the argument values for a function
// call and then perform the call itself.
// Return any value into a temporary
static int gen_funccall(ASTnode *n) {
  int i, numargs = 0;
  int *arglist = NULL;
  Type **typelist = NULL;
  Sym *func, *param;
  ASTnode *this;

  // Get the matching symbol for the function's name
  func= find_symbol(n->left->strlit);
  if (func == NULL)
    fatal("unknown function %s()\n", n->left->strlit);

  if (func->symtype != ST_FUNCTION)
    fatal("%s is not a function\n", n->left->strlit);


  // Walk the expression list to count the number of arguments
  for (this=n->right; this!=NULL; this= this->right) {
    if (this->op == A_GLUE) numargs++;
  }

  // Check the arg count vs. the function parameter count
  if (numargs != func->count)
    fatal("wrong number of arguments to %s(): %d vs. %d\n",
	n->left->strlit, numargs, func->count);

  if (numargs > 0) {
    // Allocate space to hold the types and
    // temporaries for the expressions
    arglist= (int *)malloc(numargs * sizeof(int));
    typelist = (Type **)malloc(numargs * sizeof(Type *));

    if (arglist == NULL || typelist == NULL)
      fatal("out of memory in gen_funccall()\n");

    // Walk the expression list again.
    // Check and, if needed, widen the expression's
    // type to match the parameter's type.
    // Generate the code for each expression.
    // Cache the temporary number and the type for each one.
    param= func->memb;
    for (i=0, this=n->right; this!=NULL;
			this= this->right, i++, param= param->next) {
      if (this->op == A_GLUE) {
	this->left= widen_expression(this->left, param->type);
	typelist[i]= this->left->type;
	arglist[i]= genAST(this->left);
      } else {
	this= widen_expression(this, param->type);
	typelist[i]= this->type;
	arglist[i]= genAST(this);
      }
    }
  }

  // Generate the QBE code
  cgcall(func, numargs, arglist, typelist);

  return(NOREG);
}

File: alpy/genast.py
from typing import Optional

from defs import ASTNode, ASTNodeType, DataType, SymType, fatal
from cgen import codegen, gen_label, cg_negate, cg_not, cg_invert
from strlits import get_strlit_label


class ASTCodeGenerator:
    unary_ops = {
        ASTNodeType.A_NEGATE: cg_negate,
        ASTNodeType.A_NOT: cg_not,
        ASTNodeType.A_INVERT: cg_invert,
    }

    @classmethod
    def gen_ast(cls, node: Optional[ASTNode]) -> int:
        if not node:
            return 0

        # 根据节点类型生成相应的代码
        if node.op == ASTNodeType.A_NUMLIT:
            return codegen.cg_load_lit(node.numlit, node.type)
        elif node.op == ASTNodeType.A_IDENT:
            return codegen.cg_load_var(node.sym)
        elif node.op == ASTNodeType.A_ASSIGN:
            right_temp = cls.gen_ast(node.right)
            codegen.cg_stor_var(right_temp, node.right.type, node.left.sym)
            return right_temp
        elif node.op == ASTNodeType.A_CAST:
            right_temp = cls.gen_ast(node.right)
            return codegen.cg_cast(right_temp, node.right.type, node.type)
        elif node.op in (ASTNodeType.A_NEGATE, ASTNodeType.A_NOT, ASTNodeType.A_INVERT):
            return cls.gen_unary(node)
        elif node.op in (ASTNodeType.A_ADD, ASTNodeType.A_SUBTRACT,
                         ASTNodeType.A_MULTIPLY, ASTNodeType.A_DIVIDE):
            return cls.gen_arithmetic(node)
        elif node.op in (ASTNodeType.A_AND, ASTNodeType.A_OR, ASTNodeType.A_XOR,
                         ASTNodeType.A_LSHIFT, ASTNodeType.A_RSHIFT):
            return cls.gen_logical(node)
        elif node.op in (ASTNodeType.A_EQ, ASTNodeType.A_NE, ASTNodeType.A_LE,
                         ASTNodeType.A_LT, ASTNodeType.A_GE, ASTNodeType.A_GT):
            return cls.gen_comparison(node)
        elif node.op == ASTNodeType.A_LOCAL:
            codegen.cg_add_local(node.type, node.sym)
            if node.left:
                expr_temp = cls.gen_ast(node.left)
                codegen.cg_stor_var(expr_temp, node.left.type, node.sym)
            return 0
        elif node.op == ASTNodeType.A_IF:
            return cls.gen_if(node)
        elif node.op == ASTNodeType.A_WHILE:
            return cls.gen_while(node)
        elif node.op == ASTNodeType.A_FOR:
            return cls.gen_for(node)
        elif node.op == ASTNodeType.A_FUNCCALL:
            return cls.gen_func_call(node)
        elif node.op == ASTNodeType.A_RETURN:
            expr_temp = cls.gen_ast(node.left)
            codegen.cg_ret(expr_temp)
            return expr_temp
        elif node.op == ASTNodeType.A_BLOCK:
            # 处理语句块中的所有语句
            stmt_node = node.left
            last_temp = 0
            while stmt_node:
                last_temp = cls.gen_ast(stmt_node)
                stmt_node = stmt_node.right
            return last_temp
        elif node.op in (ASTNodeType.A_PRINT, ASTNodeType.A_PRINTF):
            expr_temp = cls.gen_ast(node.right)
            # 根据类型选择合适的格式字符串
            # if is_integer(node.left.type):
            #     label = get_strlit_label("%d\n")
            # elif is_float(node.left.type):
            #     label = get_strlit_label("%f\n")
            # else:
            #     fatal(f"Unsupported type for print: {node.left.type}")
            label = get_strlit_label(node.left.strlit)
            codegen.cg_print(label, expr_temp, node.right.type)
            return 0
        elif node.op == ASTNodeType.A_GLUE:
            cls.gen_ast(node.left)
            cls.gen_ast(node.right)
            return 0
        else:
            fatal(f"Unknown AST node type: {node.op}")
            return 0

    @classmethod
    def gen_unary(cls, node: ASTNode) -> int:
        op = cls.unary_ops.get(node.op)
        rt = cls.gen_ast(node.right)
        return op(rt, node.type)

    @classmethod
    def gen_arithmetic(cls, node: ASTNode) -> int:
        lt, rt = cls.gen_ast(node.left), cls.gen_ast(node.right)
        return codegen.cg_arithmetic(node.op, lt, rt, node.type)

    @classmethod
    def gen_logical(cls, node: ASTNode) -> int:
        lt, rt = cls.gen_ast(node.left), cls.gen_ast(node.right)
        return codegen.cg_logical(node.op, lt, rt, node.type)

    @classmethod
    def gen_comparison(cls, node: ASTNode) -> int:
        lt, rt = cls.gen_ast(node.left), cls.gen_ast(node.right)
        return codegen.cg_comparison(node.op, lt, rt, node.type)

    @classmethod
    def gen_if(cls, node: ASTNode) -> int:
        cond_temp = cls.gen_ast(node.left)
        label_else = gen_label()
        label_end = gen_label()
        codegen.cg_if_false(cond_temp, label_else)
        cls.gen_ast(node.right.left)
        codegen.cg_jump(label_end)
        codegen.cg_label(label_else)
        cls.gen_ast(node.right.right)
        codegen.cg_label(label_end)
        return 0

    @classmethod
    def gen_while(cls, node: ASTNode) -> int:
        label_start = gen_label()
        label_body = gen_label()
        label_end = gen_label()
        codegen.cg_label(label_start)
        cond_temp = cls.gen_ast(node.left)
        codegen.cg_if_false(cond_temp, label_end)
        codegen.cg_label(label_body)
        cls.gen_ast(node.right)
        codegen.cg_jump(label_start)
        codegen.cg_label(label_end)
        return 0

    @classmethod
    def gen_for(cls, node: ASTNode) -> int:
        # 初始化语句
        cls.gen_ast(node.left.left)
        label_start = gen_label()
        label_body = gen_label()
        label_end = gen_label()
        codegen.cg_label(label_start)
        # 条件判断
        if node.left.right:
            cond_temp = cls.gen_ast(node.left.right)
            codegen.cg_if_false(cond_temp, label_end)
        codegen.cg_label(label_body)
        # 循环体
        cls.gen_ast(node.right.left)
        # 更新语句
        if node.right.right:
            cls.gen_ast(node.right.right)
        codegen.cg_jump(label_start)
        codegen.cg_label(label_end)
        return 0

    @classmethod
    def gen_func_call(cls, node: ASTNode) -> int:
        if not node.sym or node.sym.sym_type != SymType.SYM_FUNC:
            fatal(f"{node.sym.name if node.sym else '?'} is not a function")
        # 处理参数
        args: list[int] = []
        arg_types: list[DataType] = []
        arg_node = node.left
        while arg_node:
            arg_temp = cls.gen_ast(arg_node)
            args.append(arg_temp)
            arg_types.append(arg_node.type)
            arg_node = arg_node.right
        # 调用函数
        return codegen.cg_call(node.sym, len(args), args, arg_types)


gen_ast = ASTCodeGenerator.gen_ast
